unfurl/configurators/octodns.py

```diff
@@ -0,0 +1,194 @@
+    def _dump_current_dns_records(path: str, zone_name: str):
+        log.debug("OctoDNS configurator - downloading current DNS records")
+
+        with change_cwd(path):
+            try:
+                manager = Manager(config_file="dns/main-config.yaml")
+                manager.dump(
+                    zone_name,
+                    output_dir=f"{path}dns-dump/",
+                    lenient=False,
+                    split=False,
+                    source="target_config",
+                )
+            except Exception as e:
+                log.error("OctoDNS error: %s", e)
+
+    @staticmethod
+    def _read_current_dns_records(path: str, zone: str) -> dict:
+        records = {}
+        path = Path(path) / "dns-dump" / f"{zone}yaml"
+        if path.exists():
+            with open(path) as f:
+                yaml = YAML(typ="safe")
+                records[zone] = yaml.load(f.read())
+        return records
+
+    @staticmethod
+    def _merge_dns_records(new_zone_records: dict, old_zone_records: dict) -> dict:
+        return dict_merge(old_zone_records, new_zone_records)
+
+    @staticmethod
+    def _create_yaml_zone_files(folder: WorkFolder, records: dict):
+        for zone, content in records.items():
+            folder.write_file(content, f"dns/{zone}yaml")
+
+    def run(self, task: ConfigTask):
+        """Apply DNS configuration"""
+        op = task.configSpec.operation
+        task.logger.debug(f"OctoDNS configurator - run - {op}")
+        if op == "configure":
+            yield self._run_octodns_sync(task)  # create or update zone
+        elif op == "delete":
+            yield self._run_octodns_sync(task)  # remove zone records
+        elif op == "check":
+            yield self._run_check(task)
+        else:
+            raise NotImplementedError(f"Operation '{op}' is not allowed")
+
+    @staticmethod
+    def _run_octodns_sync(task: ConfigTask):
+        work_folder = task.set_work_folder()
+        with change_cwd(f"{work_folder.cwd}/dns"):
+            try:
+                manager = Manager(config_file="main-config.yaml")
+                manager.sync(dry_run=task.dry_run)
+                return task.done(success=True, result={"msg": "OctoDNS synced"})
+            except Exception as e:
+                log.error("OctoDNS error: %s", e)
+                return task.done(success=False, result={"msg": f"OctoDNS error: {e}"})
+
+    def _run_check(self, task: ConfigTask):
+        """Retrieves current zone data and compares with expected"""
+        work_folder = task.set_work_folder()
+        properties = self._extract_properties_from(task)
+        current_records = self._read_current_dns_records(
+            work_folder.cwd, properties.name
+        )
+
+        if current_records == properties.records:
+            return task.done(success=True, result={"msg": "DNS records in sync"})
+        else:
+            return task.done(success=False, result={"msg": "DNS records out of sync"})
```

Comment:
this should only set success to False if this operation failed with an error... if it succeeded but the records are out of sync pass status=Status.error to indicate current state of the target instance is wrong.

Reply:
Fixed